sound/soc/sof/amd/acp-ipc.c

```diff
@@ -0,0 +1,214 @@
+	    hdr->cmd == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE)) {
+		/*
+		 * memory windows are powered off before sending IPC reply,
+		 * so we can't read the mailbox for CTX_SAVE and PM_GATE
+		 * replies.
+		 */
+		reply.error = 0;
+		reply.hdr.cmd = SOF_IPC_GLB_REPLY;
+		reply.hdr.size = sizeof(reply);
+		memcpy(msg->reply_data, &reply, sizeof(reply));
+		goto out;
+	}
+	/* get IPC reply from DSP in the mailbox */
+	acp_copy_from_scratch_memory(sdev, ring_buffer, 0, &reply, sizeof(reply));
+	if (reply.error < 0) {
+		memcpy(msg->reply_data, &reply, sizeof(reply));
+		ret = reply.error;
+	} else {
+		/* reply correct size ? */
+		if (reply.hdr.size != msg->reply_size &&
+		    !(reply.hdr.cmd & SOF_IPC_GLB_PROBE)) {
+			dev_err(sdev->dev, "error: reply expected %zu got %u bytes\n",
+				msg->reply_size, reply.hdr.size);
+			ret = -EINVAL;
+		}
+		/* read the message */
+		if (msg->reply_size > 0)
+			acp_copy_from_scratch_memory(sdev, ring_buffer, 0,
+						     msg->reply_data, msg->reply_size);
+	}
+out:
+	msg->reply_error = ret;
+}
+
+irqreturn_t acp_sof_ipc_irq_thread(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+	bool ipc_irq = false;
+	int msg = 0, DspMsg = 0, DspAck = 0;
+	struct acp_dev_data *adata = sdev->pdata->hw_pdata;
+
+	DspMsg = readl(&adata->acp_sregmap->scratch_buf_info.sofDspMsgWrite);
+	if (DspMsg) {
+		snd_sof_ipc_msgs_rx(sdev);
+		acp_dsp_ipc_host_done(&adata->acp_sregmap->scratch_buf_info);
+		ipc_irq = true;
+	}
+	DspAck = readl(&adata->acp_sregmap->scratch_buf_info.sofDspAckwrite);
+	if (DspAck) {
+		spin_lock_irq(&sdev->ipc_lock);
+		/* handle immediate reply from DSP core */
+		acp_dsp_ipc_get_reply(sdev);
+		snd_sof_ipc_reply(sdev, msg);
+		/* set the done bit */
+		acp_dsp_ipc_dsp_done(&adata->acp_sregmap->scratch_buf_info);
+		spin_unlock_irq(&sdev->ipc_lock);
+		ipc_irq = true;
+	}
+	if (!ipc_irq) {
+       /*
+	* This interrupt is not shared so no need to return IRQ_NONE.
+	*/
+		dev_dbg_ratelimited(sdev->dev,
+				"nothing to do in IPC IRQ thread\n");
+	}
+	return IRQ_HANDLED;
+}
+EXPORT_SYMBOL_NS(acp_sof_ipc_irq_thread, SND_SOC_SOF_AMD_COMMON);
+/*
+ * This function is called when DSP has some thing to send
+ * data to Host.So the direction is DSP->Host
+ * @sdev: sof  device.
+ * @stream : DSP stream..
+ * @p:message to host
+ * @sz : size of msg.
+ */
+int acp_sof_ipc_msg_data(struct snd_sof_dev *sdev,
+			  struct snd_pcm_substream *substream,
+			  void *p, size_t sz)
+{
+	struct acp_dev_data *adata = sdev->pdata->hw_pdata;
+	unsigned char *ring_buffer = adata->acp_sregmap->scratch_buf_info.sofOutBoxBuffer;
+
+	if (!substream || !sdev->stream_box.size)
+		acp_copy_from_scratch_memory(sdev, ring_buffer, 0, p, sz);
+
+	return 0;
+}
+EXPORT_SYMBOL_NS(acp_sof_ipc_msg_data, SND_SOC_SOF_AMD_COMMON);
+
+
+int acp_sof_ipc_pcm_params(struct snd_sof_dev *sdev,
+			struct snd_pcm_substream *substream,
+			const struct sof_ipc_pcm_params_reply *reply)
+{
+	return 0;
```

Comment:
is this a TODO?

Reply:
indeed this is a bit strange, hw_params are rather helpful....

Reply:
Agreed this is helpful and in our TODO list , will update comments in file. Since this one is mandatory ops we are simply returning 0, in near future we will implement properly.